src/lib/core/exception.cc

```diff
@@ -118,6 +136,34 @@ Exception::Exception(const struct type_info *type_arg, const char *file,
 		     exception_log, type_arg, file, line);
 }
 
+Exception *Exception::deep_copy() const
+{
+	Exception *head = NULL;
+	try {
+		head = dup();
+		head->cause = cause;
```

Comment:
You should set `head->cause` only to duplicated cause. This line of code:
1. Meaningless. Head is not intended to hold this pointer.
2. Can confuse the reader of that code.
3. Leads to severe bug. Imagine the next `dup` will throw, you'll catch it and start to cleanup. You will delete `cause` in current Exception object, leaving broken pointer.

Reply:
Thank you for the detailed explanation.
Seems to be fixed.